raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -167,16 +229,35 @@ public boolean updateLocalEndOffset(LogOffsetMetadata endOffsetMetadata) {
 
     private static class ReplicaState implements Comparable<ReplicaState> {
         final int nodeId;
-        boolean hasEndorsedLeader;
+
         Optional<LogOffsetMetadata> endOffset;
+
         OptionalLong lastFetchTimestamp;
+        long lastCaughtUpTimeMs = 0L;
+        long lastFetchLeaderLogEndOffset;
 
         public ReplicaState(int nodeId,
-                            boolean hasEndorsedLeader) {
+                            long lastLeaderLogEndOffset) {
```

Comment:
See my other comment: I think we do not need to keep track of the last leader log end offset.